homeassistant/components/heos/services.py

```diff
@@ -24,27 +32,52 @@
 
 HEOS_SIGN_OUT_SCHEMA = vol.Schema({})
 
+HEOS_JOIN_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_GROUP_LEADER): cv.entity_id,
+        vol.Required(ATTR_ENTITY_ID): cv.comp_entity_ids,
+    }
+)
+
+HEOS_UNJOIN_SCHEMA = vol.Schema(
+    {vol.Optional(ATTR_ENTITY_ID, default=[]): cv.entity_ids}
+)
+
 
 def register(hass: HomeAssistantType, controller: Heos):
     """Register HEOS services."""
     hass.services.async_register(
-        DOMAIN,
+        HEOS_DOMAIN,
         SERVICE_SIGN_IN,
         functools.partial(_sign_in_handler, controller),
         schema=HEOS_SIGN_IN_SCHEMA,
     )
     hass.services.async_register(
-        DOMAIN,
+        HEOS_DOMAIN,
         SERVICE_SIGN_OUT,
         functools.partial(_sign_out_handler, controller),
         schema=HEOS_SIGN_OUT_SCHEMA,
     )
+    hass.services.async_register(
+        HEOS_DOMAIN,
+        SERVICE_JOIN,
```

Comment:
@andrewsayre : Your suggestion makes perfect sense. I guess you are talking about an `async_register_entity_service` call in `media_player`, just like this:

```python
component.async_register_entity_service(
    SERVICE_JOIN,
    {vol.Required(ATTR_ENTITY_ID): list},
    "async_join",
    [SUPPORT_GROUPING],
)
```

This would group all the players specified in `entity_id` together, hiding the leader concept. While this is a very nice and clean API (I really like it) it cannot be implemented like this because:

- This definition would call `async_join` on all players in `entity_id`
- The method `async_join` would not receive any arguments

I hope you see where my problem is here. We could implement it this way:

```python
component.async_register_entity_service(
    SERVICE_JOIN,
    {vol.Required(ATTR_GROUP_MEMBERS): list},
    "async_join",
    [SUPPORT_GROUPING],
)
```

Which means we'll end up with two arguments in the service call again: `entity_id` and `group_members`. For HEOS this means that `entity_id` will always be the leader and `group_membets` will be ... the group members. Would this be ok?

Reply:
Ok, I decided split `entity_id` from `group_members` in the service call. This way it can be implemented as an entity service.